HPCC-18408 Create an interface to hide dali esdl access details

mayx opened the pull request:
- Add interface IEsdlStore to abstract esdl dali operations
- Move dali related operations from esdl binding and ws_esdlconfig to
  a class that implements IEsdlStore. Most of the implementation function
  remain unchanged during the move.
- Add IEsdlSubscription and IEsdlListener interfaces to abstract
  subscription and notify operations with dali

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

mayx commented:
@afishbeck @rpastrana Please code review...

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: fe0a13f08226a3ee0c61f356fdd3922d9887081a
Empty table
Build: success
Build: success
ECL Watch:  Rebuilding Site
Empty table
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting mydali ...           \x1b[32m [   OK    ]  \x1b[0m 
Starting mydfuserver ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclagent ...       \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclccserver ...    \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclscheduler ...   \x1b[32m [   OK    ]  \x1b[0m 
Starting myesp ...            \x1b[31m [ FAILED  ]  \x1b[0m 
Starting myroxie ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mysasha ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mythor ...           \x1b[32m [   OK    ]  \x1b[0m 
HPCC status:
mydafilesrv     ( pid    24443 ) is running ...
mydali          ( pid    25923 ) is running ...
mydfuserver     ( pid    27663 ) is running ...
myeclagent      ( pid    29415 ) is running ...
myeclccserver   ( pid    31112 ) is running ...
myeclscheduler  ( pid      454 ) is running ...
myesp           is stopped,
myroxie         ( pid     4209 ) is running ...
mysasha         ( pid     6192 ) is running ...
mythor          ( pid     8168 ) is running with 1 slave process(es) ...
myesp           is stopped,
Archive HPCC logs.
Removing RPM
Loaded plugins: fastestmirror
Resolving Dependencies
--> Running transaction check
---> Package hpccsystems-platform.x86_64 0:6.5.0-trunk0 will be erased
--> Finished Dependency Resolution

Dependencies Resolved

================================================================================
 Package                   Arch        Version             Repository      Size
================================================================================
Removing:
 hpccsystems-platform      x86_64      6.5.0-trunk0        installed      158 M

Transaction Summary
================================================================================
Remove  1 Package

Installed size: 158 M
Downloading packages:
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
Stopping mythor...            \x1b[32m [   OK    ]  \x1b[0m 
Stopping mysasha...           \x1b[32m [   OK    ]  \x1b[0m 
Stopping myroxie...           \x1b[32m [   OK    ]  \x1b[0m 
Stopping myesp...             \x1b[32m [   OK    ]  \x1b[0m 
Stopping myeclscheduler...    \x1b[32m [   OK    ]  \x1b[0m 
Stopping myeclccserver...     \x1b[32m [   OK    ]  \x1b[0m 
Stopping myeclagent...        \x1b[32m [   OK    ]  \x1b[0m 
Stopping mydfuserver...       \x1b[32m [   OK    ]  \x1b[0m 
Stopping mydali...            \x1b[32m [   OK    ]  \x1b[0m 
Service dafilesrv, mydafilesrv is still running.
To stop it, run service dafilesrv stop.
Stopping mydafilesrv...       \x1b[32m [   OK    ]  \x1b[0m 

Removing Symlink /etc/init.d/dafilesrv                        ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/dfuplus                             ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/ecl                                 ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/eclcc                               ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/eclplus                             ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/soapplus                            ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/testsocket                          ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/wuget                               ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Sym
 ... ( comment is too long, 110 lines, 5724 bytes, truncated) ... 
ving Directory - /var/lock/HPCCSystems
Removing Directory - /var/log/HPCCSystems
Removing Directory - /var/run/HPCCSystems
Removing Directory - /var/lib/HPCCSystems
Removing user - hpcc
HPCC Uninstall: OK
Finish at: 2017-11-16_05-02-50
All done.

```

richardkchapman commented:
@mayx can you check into why smoketest failed

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 354a55dee41ce7dcf81417ccec56aa117cb2184c
Empty table
Build: success
Build: success
ECL Watch:  Rebuilding Site
Empty table
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |117 |117 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |745 |745 |0 |
|test (thor) |640 |640 |0 |
|test (roxie) |773 |773 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

rpastrana commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -0,0 +1,865 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2013 HPCC Systems®.
```
update date on new files.

rpastrana commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -0,0 +1,865 @@
+#include <memory>
+
+static const char* ESDL_DEFS_ROOT_PATH="/ESDL/Definitions/";
+static const char* ESDL_DEF_PATH="/ESDL/Definitions/Definition";
```
It's my understanding that the major performance gain would come from changing the Definitions to be uniquely identifiable based on the element name... 
ie
/ESDL/Definitions/Definition @id=mathservice.1 would turn to 
/ESDL/Definitions/mathservice.1

two points... 
Although that is the one performance improvement suggested by @jakesmith, I'm not sure if the performance improvement potential is high enough to pursue or not, but I do believe that's what Jake pointed out.
Also, I'm not sure if now we need to set a nomenclature rule for the definition names. in other words, I don't know if allowing the Definitions to be named freely by the developer is ok or not, or if we need to come up w/ a standard naming convention like the workunits... /Workunits/W<timestamp>

mayx replied:
@rpastrana the scope of this issue is to create the interfaces and encapsulate the Dali operations. There's another issue for the performance improvement(HPCC-18492). The 2 concerns you pointed out are very good points, I'll copy over the comments to the other issue, and we'll discuss further.

rpastrana commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -0,0 +1,865 @@
+    }
+};
+
+class CEsdlSDSSubscription : implements IEsdlSubscription, public CInterface, implements ISDSSubscription
```
The other improvement @jakesmith mentioned was to minimize/eliminate the dali subscriptions all together. Instead of relying on the dali notifications, wsesdlconfig could communicate the esdl def/binding change directly to the appropriate DESDL binding/service.

mayx replied:
@rpastrana again this issue is simply create the interface and set up the stage for other improvements. I've created another task for the subscription improvement: https://track.hpccsystems.com/browse/HPCC-18754, as a subtask of 18153.

jakesmith replied:
@mayx - it may be beyond the scope of this PR, but from a brief look at it, you're new interfaces are tied to a subscription / notify paradigm.
Is the intention to then remove these at later point, when the updates are pushed without a subscription?

mayx replied:
@jakesmith The new subscription interface is to isolate the components that will act on the changes(consumer), and the generic component that will receive the changes(producer). Currently the producer uses Dali subscription and notification, but it can also be a service that receives pushes. The Subscription is simply a bridge between the consumer and the producer for them to communicate asynchronously. Some function parameters (for example espprocess and binding) may change a bit later with other planned changes, but the paradigm should work with pushing notifications.

rpastrana reviewed (changes requested):
@mayx a couple of minor comments, and a couple of architectural concerns regarding the dali interaction improvements suggested by @jakesmith. We can meet and discuss offline.

mayx commented:
@rpastrana @afishbeck changes made based on comments, please review again...

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 557613b8ecd90e4e75c5464b50fe03e98a37bdbe
Empty table
Build: success
Build: success
ECL Watch:  Rebuilding Site
Empty table
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |117 |117 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |745 |745 |0 |
|test (thor) |640 |640 |0 |
|test (roxie) |773 |773 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented:
@mayx - no other comments.
@afishbeck - may be good if you take a final review.

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+interface IEsdlStore : public IInterface
+{
+    virtual void fetchDefinition(const char* id, StringBuffer& esxdl) = 0;
+    virtual void fetchDefinition(const char* esdlServiceName, unsigned targetVersion, StringBuffer& esxdl) = 0;
```
It also might be less clutter just to always let the caller build the id, rather than having 2 methods, but I could be convinced either way.

mayx replied:
The second fetch function removed.

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+    virtual void fetchDefinition(const char* id, StringBuffer& esxdl) = 0;
+    virtual void fetchDefinition(const char* esdlServiceName, unsigned targetVersion, StringBuffer& esxdl) = 0;
+    virtual void fetchDefinitionByNameOnly(const char * name, StringBuffer & def) = 0;
+    virtual IPropertyTree* fetchBinding(const char* espProcess, const char* bindingName) = 0;
```
This whole binding term gets pretty confusing because we have ESP/ENV_Configured/static bindings, and ESDL/DALI_ESDL_Configured/Virtual bindings.  Wish I'd called the ESDL one something else.  Anyway perhaps for this method.. fetchESDLBinding(const char *espProcess, const char *espStaticBinding);  Just to be ultra clear.

Some of your other issues may change this even further because we'll really tie ESDLBindings to a port, rather than a preconfigured ESPStaticBinding.  Eventually maybe we'll get the naming down so it's not so confusing.

mayx replied:
Changed argument name to espStaticBinding

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+    virtual IPropertyTree* fetchBinding(const char* espProcess, const char* bindingName) = 0;
+    virtual bool isDefinitionBound(const char * esdldefid) = 0;
+    virtual bool isDefinitionBound(const char * esdldefname, int version) = 0;
+    virtual IPropertyTree* getDefinitionRegistry(const char * wsEclId, bool readonly) = 0;
```
wsEclId is not a good name... but maybe that parameter is not even used?  Returning the registry as a property tree isn't a good idea... when we move this stuff from DALI it may have to change.  It works for DALI because DALI is already like a giant remote property tree.

mayx replied:
Removed argument wsEclId as it's not being used in the function.

mayx replied:
I think we can keep the return type as IPropertyTree for now, being aware that this may need to change.

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+    virtual bool isDefinitionBound(const char * esdldefid) = 0;
+    virtual bool isDefinitionBound(const char * esdldefname, int version) = 0;
+    virtual IPropertyTree* getDefinitionRegistry(const char * wsEclId, bool readonly) = 0;
+    virtual bool existsDefinition(const char * definitionid) = 0;
```
Just me but for some reason I don't like the name in that order.  Maybe checkDefinitionExists() or definitionExists()?  But no big deal.

mayx replied:
Changed function name to definitionExists.

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+    virtual bool isDefinitionBound(const char * esdldefname, int version) = 0;
+    virtual IPropertyTree* getDefinitionRegistry(const char * wsEclId, bool readonly) = 0;
+    virtual bool existsDefinition(const char * definitionid) = 0;
+    virtual bool existsDefinition(const char * servicename, unsigned ver) = 0;
```
All of these helper functions that separate version seem like clutter, but can be convinced otherwise.

mayx replied:
Removed existsDefinition that takes name and ver. Removed isDefinitionBound from the interface.

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+    virtual IPropertyTree* getDefinitionRegistry(const char * wsEclId, bool readonly) = 0;
+    virtual bool existsDefinition(const char * definitionid) = 0;
+    virtual bool existsDefinition(const char * servicename, unsigned ver) = 0;
+    virtual bool existsMethodDef(const char * esdlDefinitionName, unsigned ver, StringBuffer & esdlServiceName, const char * methodName) = 0;
```
checkMethodDefinitionExists, or isMethodDefined?  Need to be consistent on the variable name of "definitionName".  But maybe instead of name and version we use, "definitionId"?

mayx replied:
Changed name to isMethodDefined, and changed to use defintionId.

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+    virtual bool existsDefinition(const char * definitionid) = 0;
+    virtual bool existsDefinition(const char * servicename, unsigned ver) = 0;
+    virtual bool existsMethodDef(const char * esdlDefinitionName, unsigned ver, StringBuffer & esdlServiceName, const char * methodName) = 0;
+    virtual void addDefinition(IPropertyTree * queryRegistry, const char * name, IPropertyTree *definitionInfo, StringBuffer &newId, unsigned &newSeq, const char *userid, bool deleteprev) = 0;
```
Change queryRegistry to definitionRegistry (it's left over from this being based on publishing queries). "defintionName"

mayx replied:
Changed the argument names as suggested.

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+    virtual bool existsDefinition(const char * servicename, unsigned ver) = 0;
+    virtual bool existsMethodDef(const char * esdlDefinitionName, unsigned ver, StringBuffer & esdlServiceName, const char * methodName) = 0;
+    virtual void addDefinition(IPropertyTree * queryRegistry, const char * name, IPropertyTree *definitionInfo, StringBuffer &newId, unsigned &newSeq, const char *userid, bool deleteprev) = 0;
+    virtual int publishMethod(const char * espProcName, const char * espBindingName, const char * srvDefId, const char * methodName, IPropertyTree * attributesTree, bool overwrite, StringBuffer & message) = 0;
```
Perhaps "configureMethod".. all this overloading of the word "publish" gets confusing.  Maybe attributesTree should be config or configTree.

mayx replied:
Change made per suggested.

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+    virtual bool existsMethodDef(const char * esdlDefinitionName, unsigned ver, StringBuffer & esdlServiceName, const char * methodName) = 0;
+    virtual void addDefinition(IPropertyTree * queryRegistry, const char * name, IPropertyTree *definitionInfo, StringBuffer &newId, unsigned &newSeq, const char *userid, bool deleteprev) = 0;
+    virtual int publishMethod(const char * espProcName, const char * espBindingName, const char * srvDefId, const char * methodName, IPropertyTree * attributesTree, bool overwrite, StringBuffer & message) = 0;
+    virtual int publishBinding(const char * bindingName, IPropertyTree * methodsConfig, const char * espProcName, const char * espPort, const char * esdlDefinitionName,
```
Again avoid publish except at the highest level.  Maybe "bindService" or "bindESDLService".  Maybe combine definitonName and definitionVersion into one "defintionId" for consistency?

mayx replied:
Changed method name to bindService, and replace name version with definitionId.

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+    virtual int publishMethod(const char * espProcName, const char * espBindingName, const char * srvDefId, const char * methodName, IPropertyTree * attributesTree, bool overwrite, StringBuffer & message) = 0;
+    virtual int publishBinding(const char * bindingName, IPropertyTree * methodsConfig, const char * espProcName, const char * espPort, const char * esdlDefinitionName,
+                               int esdlDefinitionVersion, const char * esdlServiceName, StringBuffer & message, bool overwrite, const char * user) = 0;
+    virtual IPropertyTree* getBindingTree(const char * espProcName, const char * espBindingName, StringBuffer & msg) = 0;
```
This is another one that may evolve as you work on other issues.  We really need to support more than one esdl service being bound per port.  Also may evolve when not using DALI as back end.

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+    virtual int publishBinding(const char * bindingName, IPropertyTree * methodsConfig, const char * espProcName, const char * espPort, const char * esdlDefinitionName,
+                               int esdlDefinitionVersion, const char * esdlServiceName, StringBuffer & message, bool overwrite, const char * user) = 0;
+    virtual IPropertyTree* getBindingTree(const char * espProcName, const char * espBindingName, StringBuffer & msg) = 0;
+    virtual bool deleteDefinition(const char* definitionId, StringBuffer& errmsg, StringBuffer& defxml) = 0;
```
For an abstract interface seems like returning what you are deleting might not always be needed... so could make it optional... StringBuffer *defxml and only populate it if the pointer is not nullptr.

mayx replied:
Changed to StringBuffer*

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+                               int esdlDefinitionVersion, const char * esdlServiceName, StringBuffer & message, bool overwrite, const char * user) = 0;
+    virtual IPropertyTree* getBindingTree(const char * espProcName, const char * espBindingName, StringBuffer & msg) = 0;
+    virtual bool deleteDefinition(const char* definitionId, StringBuffer& errmsg, StringBuffer& defxml) = 0;
+    virtual bool deleteBinding(const char* bindingId, StringBuffer& errmsg, StringBuffer& bindingxml) = 0;
```
Again, optional bindingxml returned.

mayx replied:
Changed to StringBuffer*

afishbeck commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+    virtual IPropertyTree* getBindingTree(const char * espProcName, const char * espBindingName, StringBuffer & msg) = 0;
+    virtual bool deleteDefinition(const char* definitionId, StringBuffer& errmsg, StringBuffer& defxml) = 0;
+    virtual bool deleteBinding(const char* bindingId, StringBuffer& errmsg, StringBuffer& bindingxml) = 0;
+    virtual IPropertyTree* getDefinitions() = 0;
```
These will probably need abstracting later, when moving away from DALI.

afishbeck reviewed (changes requested):
Comments mostly on interface level for now.

richardkchapman commented:
@mayx Back to you

mayx commented on esp/services/esdl_svc_engine/esdl_store.cpp:
```diff
@@ -0,0 +1,865 @@
+
+esdl_engine_decl IEsdlStore* createEsdlCentralStore()
+{
+    return new CEsdlSDSStore;
```
Makes sense, will make the change.

mayx commented on esp/services/esdl_svc_engine/esdl_store.hpp:
```diff
@@ -0,0 +1,82 @@
+{
+    virtual void fetchDefinition(const char* id, StringBuffer& esxdl) = 0;
+    virtual void fetchDefinition(const char* esdlServiceName, unsigned targetVersion, StringBuffer& esxdl) = 0;
+    virtual void fetchDefinitionByNameOnly(const char * name, StringBuffer & def) = 0;
```
Function name changed to fetchLatestDefinition.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 776e439ec686600d2d587611988ff3b6f0b9abdd
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |92 |92 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |735 |735 |0 |
|test (thor) |648 |648 |0 |
|test (roxie) |764 |764 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   24 sec (00:00:24) |  169 sec (00:02:49) |   59 sec (00:00:59) |    8 sec (00:00:08) |   55 sec (00:00:55) | 1108 sec (00:18:28) |   39 sec (00:00:39) | 1462 sec (00:24:22) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 218a5efeac269a37789fa6643831b07374dbf203
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |92 |92 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |735 |735 |0 |
|test (thor) |648 |648 |0 |
|test (roxie) |764 |764 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  172 sec (00:02:52) |   60 sec (00:01:00) |    9 sec (00:00:09) |   55 sec (00:00:55) | 1080 sec (00:18:00) |   38 sec (00:00:38) | 1422 sec (00:23:42) |

mayx commented:
@afishbeck changes made as suggested, please review again.

afishbeck reviewed (approved):
@mayx looks fine

mayx commented:
@rpastrana I resolved the merging conflicts and rebased with master, please do a final review...

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -67,75 +67,6 @@ bool trimXPathToParentSDSElement(const char *element, const char * xpath, String
-    return NULL;
-}
-
 IPropertyTree * fetchESDLBindingFromStateFile(const char *process, const char *bindingName, const char * stateFileName)
```
why don't we move this method to your new class as well?

mayx replied:
For this pass I'd like to leave the state file out and only deal with "central" repository.

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1096,6 +933,7 @@ EsdlBindingImpl::EsdlBindingImpl()
 
 EsdlBindingImpl::EsdlBindingImpl(IPropertyTree* cfg, const char *binding,  const char *process) : CHttpSoapBinding(cfg, binding, process)
 {
+    m_pCentralStore.setown(createEsdlCentralStore());
```
does createEsdlCentralStore return a singleton? if not, is there any reason why it can't be?

mayx replied:
Yes it's already a singleton.

rpastrana commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1117,25 +955,14 @@ EsdlBindingImpl::EsdlBindingImpl(IPropertyTree* cfg, const char *binding,  const
             DBGLOG("ESDL Binding: Could not fetch ESDL binding %s for ESP Process %s", binding, process);
 
-        //Subscribe to Dali anyway
         DBGLOG("ESDL Binding %s is subscribing to all /ESDL/Bindings/Binding dali changes", binding);
```
maybe this log entry should be in the centralstore?

mayx replied:
Removed this log and enhanced corresponding log in central store.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 6710fb7cb27e5ccac5508d76d9cb85b1f3bb92d8
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |82 |82 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |650 |650 |0 |
|test (roxie) |768 |768 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    5 sec (00:00:05) |  178 sec (00:02:58) |   56 sec (00:00:56) |    0 sec (00:00:00) |   36 sec (00:00:36) | 1125 sec (00:18:45) |   45 sec (00:00:45) | 1445 sec (00:24:05) |

rpastrana reviewed (commented):
@mayx since most of this had already been reviewed, I focused on ensuring proper integration of  HPCC-18881
and HPCC-18905. I confirmed the former, but I didn't see where the latter was integrated. Please confirm.
Also, there were a couple of other comments in line.

mayx committed:
HPCC-18408 Create an interface to hide dali esdl access details and xpaths

- Add interface IEsdlStore to abstract esdl dali operations
- Move dali related operations from esdl binding and ws_esdlconfig to
  a class that implements IEsdlStore. Most of the implementation function
  remain unchanged during the move.
- Add IEsdlSubscription and IEsdlListener interfaces to abstract
  subscription and notify operations with dali

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

mayx commented:
@rpastrana please review again. It seems that 18905 was merged in before, or same changes were made in master, that's why the difference didn't show.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 53c7a3ba56d0c550a92852591a7a1a32b15ee1c1
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |82 |82 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |650 |650 |0 |
|test (roxie) |768 |768 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    6 sec (00:00:06) |  183 sec (00:03:03) |   56 sec (00:00:56) |    5 sec (00:00:05) |   36 sec (00:00:36) | 1073 sec (00:17:53) |   32 sec (00:00:32) | 1391 sec (00:23:11) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 96dbe3b44c756d3d6081b35230b3be416f717991
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |761 |761 |0 |
|test (thor) |675 |675 |0 |
|test (roxie) |788 |788 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    6 sec (00:00:06) |  180 sec (00:03:00) |   56 sec (00:00:56) |    6 sec (00:00:06) |   35 sec (00:00:35) | 1163 sec (00:19:23) |   32 sec (00:00:32) | 1478 sec (00:24:38) |

rpastrana reviewed (approved):
@mayx looks fine
